Improve type docs

dmolineus opened the pull request:
This pull request improves some type documentations. When working with static analysis tools like psalm in a strict level Contao type docs are a bit cumbersome on many places.

dmolineus committed:
Fix existing code style violations

dmolineus committed:
Fix docs of extensions parameter

The method gets a list of extensions configurations and has to return it

dmolineus committed:
Improve configuration of replace and load after settings

m-vo commented:
Just out of interest:

 `list<x>` means `[a_of_x, b_of_x, …]` which, in PHP, is non-distinguishable from `[0 => a_of_x, 1 => b_of_x. …]`, i.e. a  associative array with zero based numeric keys with no gaps. So is using `list<x>` over `array<x>` or `array<int, x>` purely there for semantics or does it guarantee that keys are indeed canonical?

discordier commented:
> does it guarantee that keys are indeed canonical?

Yep, that's the definition (according to psalm), `array_values` returns a list of elements.
See also https://psalm.dev/docs/annotating_code/type_syntax/array_types/

fritzmg commented on src/Bundle/Config/ConfigResolver.php:
```diff
@@ -169,7 +169,7 @@ private function normalizeLoadingOrder(array $loadingOrder, array $replace): arr
     }
 
-    private function replaceBundleNames(array &$loadAfter, array $replace): void
+    private function replaceBundleNames(array & $loadAfter, array $replace): void
```
Did you do this on purpose or was it the CS fixer?

dmolineus replied:
That's part of [7addeac](https://github.com/contao/manager-plugin/pull/43/commits/7addeac912d4f5a155a469999c85dc3afd3512cb) which just contains automatic fixes of the cs fixer

fritzmg replied:
Yes, I think that is unintended though.

dmolineus replied:
I ran `composer bin ecs update` and running the cs fixer again undo the changes ([db09287](https://github.com/contao/manager-plugin/pull/43/commits/db09287998ecf5f1a6fe783f6d0485c93cb28c0f)).

dmolineus commented:
Is there anything missing or should be improved here?

leofeyer reviewed (changes requested):
`::class` on objects is only supported as of PHP 8.0, however, the plugin supports PHP 7.1+ as well. We have to keep the `\get_class()` calls therefore.

aschempp commented:
ping @dmolineus

dmolineus committed:
Run composer bin ecs update and fix code style again

dmolineus commented:
> `::class` on objects is only supported as of PHP 8.0, however, the plugin supports PHP 7.1+ as well. We have to keep the `\get_class()` calls therefore.

I adjusted the coding standard and removed the `::class` usages

leofeyer commented on src/Config/ExtensionPluginInterface.php:
```diff
@@ -17,9 +17,10 @@ interface ExtensionPluginInterface
+     * @param list<array<string,mixed>> $extensionConfigs
      *
-     * @return array<string,mixed>
+     * @return list<array<string,mixed>>
```
```suggestion
     * @return list<string,mixed>
```

Otherwise it is one array level too much, isn‘t it?

ausi replied:
`list<string,mixed>` is invalid, because `list` can only have one type inside `<…>` (no type for the key as this is always `int` for lists)

I think `list<array<string,mixed>>` is correct because it seems to describe a structure like this:

```php
[
    [
        'string1' => …mixed…,
        'string2' => …mixed…,
    ],
    [
        'string3' => …mixed…,
    ],
]
```

leofeyer committed:
CS

leofeyer committed:
Merge branch 'main' into hotfix/type-docs

leofeyer commented:
Thank you @dmolineus.
